Port makeFeaturesCompatible from python to C++

elpaso opened the pull request:
Part of the in-place processing editing, with full test coverage.

elpaso committed:
Port makeFeaturesCompatible to C++

as: QgsVectorLayerUtils::makeFeaturesCompatible

With tests.

elpaso committed:
Revert const on createFeature

just because it would be an API break

elpaso committed:
Const correctness (because it's not an API break)

nyalldawson commented on src/core/qgsvectorlayerutils.cpp:
```diff
@@ -522,6 +523,135 @@ std::unique_ptr<QgsVectorLayerFeatureSource> QgsVectorLayerUtils::getFeatureSour
+  QgsFeatureList resultFeatures;
+  for ( const QgsFeature &f : features )
+  {
+    QgsFeature newF( f );
```
Could you pull this loop out into its own function (makeFeatureCompatible)? I could see that calling this on an individual feature would be handy too.

nyalldawson commented on src/core/qgsvectorlayerutils.cpp:
```diff
@@ -522,6 +523,135 @@ std::unique_ptr<QgsVectorLayerFeatureSource> QgsVectorLayerUtils::getFeatureSour
+        for ( int i = 0; i < parts->partCount( ); i++ )
+        {
+          QgsGeometry g( parts->geometryN( i )->clone() );
+          QgsAttributeMap attrMap;
```
This should be pulled out of the part loop, `attrMap` can be reused for every part.

nyalldawson commented:
Note that https://github.com/qgis/QGIS/pull/7998 also includes the attribute commit, along with a port of the unit tests from the processing in-place tests

elpaso commented:
Then better merge #7998 first, then I'll deal with conflicts if any.

nyalldawson commented:
Ok - I'm gonna leave #7998 open for comments for a while first, it's important to get right

elpaso committed:
makeFeatureCompatible on a single input feature

elpaso committed:
Typo

m-kuhn commented on src/core/qgsvectorlayerutils.h:
```diff
@@ -167,6 +167,56 @@ class CORE_EXPORT QgsVectorLayerUtils
+     *
+     * \since QGIS 3.4
+     */
+    static const QgsFeatureList makeFeatureCompatible( const QgsFeature &feature, const QgsVectorLayer &layer );
```
can you make these pointers `QgsVectorLayer *layer`?

m-kuhn commented on src/core/qgsvectorlayerutils.cpp:
```diff
@@ -522,6 +523,147 @@ std::unique_ptr<QgsVectorLayerFeatureSource> QgsVectorLayerUtils::getFeatureSour
+      attributes.reserve( fields.count() );
+      for ( int i = feature.attributes().count(); i < fields.count(); ++i )
+      {
+        attributes.append( QVariant( fields.at( i ).type() ) );
```
Sidenote: I think the `QVariant` here is not required.

m-kuhn commented on src/core/qgsvectorlayerutils.cpp:
```diff
@@ -522,6 +523,147 @@ std::unique_ptr<QgsVectorLayerFeatureSource> QgsVectorLayerUtils::getFeatureSour
+  QgsFeatureList resultFeatures;
+  for ( const QgsFeature &f : features )
+  {
+    for ( const auto &_f : makeFeatureCompatible( f, layer ) )
```
needs to take a const copy of this

elpaso replied:
even if `makeFeatureCompatible` already returns a const? I'm pretty sure it won't detach, but maybe it is compiler-specific.

m-kuhn replied:
I think we had this discussion before and [that was the result](https://www.kdab.com/goodbye-q_foreach/#comment-143577).

m-kuhn commented on src/core/qgsvectorlayerutils.h:
```diff
@@ -167,6 +167,56 @@ class CORE_EXPORT QgsVectorLayerUtils
+     * Matches the attributes in \a feature to the specified \a fields.
+     *
+     * This causes the attributes contained within the given \a feature to be rearranged (or in
+     * some cases dropped) in order to match the fields and order indicated by \a fields.
```
Should there be a note included about the behavior difference if `feature.fields().empty()` vs. a feature with fields set?

nyalldawson replied:
Fixed in #7998

m-kuhn commented on src/core/qgsvectorlayerutils.h:
```diff
@@ -167,6 +167,56 @@ class CORE_EXPORT QgsVectorLayerUtils
+     *
+     * This function returns a new list of transformed features compatible with the input
+     * layer, note that the number of features returned might be greater than the number
+     * of input featurers.
```
in case multipart need to be converted to single part features?

elpaso replied:
Yes, exactly

DelazJ replied:
typo on "features"

m-kuhn replied:
The question was more "does one need to read the code to find out or should it be part of the API docs" ;)

m-kuhn commented:
Is looking good, just a couple of minor comments here

elpaso committed:
Some minor changes

- QgsVectorLayer *
- make a const copy of a const container
- typo and comments

m-kuhn commented:
Thanks, looking much better. So the final request would be to remove the `const` from the return value as per [the discussion linked before](https://www.kdab.com/goodbye-q_foreach/#comment-143577).

elpaso commented:
@m-kuhn 

Sorry, that discussion is a bit cryptic to me: does it mean that returning const is "bad" because if you do so you cannot use move semantic (explicitely or  through RVO) ?

It does not seem to say that iterating over a returned const container will trigger a copy, does it?

So, if I got that right, what you are saying is that:
- better do not return a const or move semantic will break
- if you do not return a const then you need to make a const copy (that is obvious)
- if you do return a const then making a const copy is not necessary

Perhaps I won't bother you again if I finally understand where is the issue :)

m-kuhn commented:
If I am reading that right (and it's also a bit more than the usual amount of C++ I am used to digest)

 * it (probably?) works to iterate over a const return value without detaching. It does neither confirm nor deny explicitly, but you did some tests previous to this discussion
 * but it's not something they will do on Qt side, because it breaks move semantics

What this means for us:

 * we should also not do it because in most cases Qt does things right and I'm happy to follow their lead
 * additional consideration, for reviewers it's much easier to spot unintended detaches if one does not have to also lookup the signature of the called method

elpaso commented:
@m-kuhn thanks, fair enough, I wonder if what Marc said is that being move semantic broken, there would be a copy anyway, when you take the temporary to iterate over (even if my previous tests suggested that it's not the case).

m-kuhn commented:
I think it's the contrary. While it may work to prevent the compiler from detaching in this case (by taking a non-const copy), in any other case where you would want to apply move semantics on the returned value for whatever purpose this will not work. [See also "Can a C++ compiler perform RVO for a const return value?"](https://stackoverflow.com/questions/14429988/can-a-c-compiler-perform-rvo-for-a-const-return-value)

elpaso committed:
Drop const on makeFeatureComptible returned values
